File: src/regression/scala_liblinear/Regression.scala
import scala.collection.mutable.ArraySeq

import java.io.File;
import java.io.IOException;
import de.bwaldvogel.liblinear.Feature;
import de.bwaldvogel.liblinear.FeatureNode;
import de.bwaldvogel.liblinear.Linear;
import de.bwaldvogel.liblinear.Model;
import de.bwaldvogel.liblinear.Parameter;
import de.bwaldvogel.liblinear.Problem;
import de.bwaldvogel.liblinear.SolverType;

object Regression {

	/**
 	* Train a model on X with labels Y 
 	*/
	def train(dm : DataMatrix) = { 
		val X : ArraySeq[Array[FeatureNode]] = dm.getX();
		val Y : ArraySeq[Double] = dm.getY();
		
/*		val problem : Problem = new Problem();
		problem.x = X;  // feature nodes
		problem.y = Y;  // target values
		problem.l = X.length; // number of training examples
		problem.n = X(0).length; // number of features

		val solver : SolverType = SolverType.L2R_LR; 
		val C : Double = 1.0;    // cost of constraints violation
		val eps : Double = 0.01; // stopping criteria

		val parameter : Parameter = new Parameter(solver, C, eps);
		Linear.train(problem, parameter);*/
	}

	/**
 	* Test model on X with known labels Y 
 	*/
	def test(model : Model, dm : DataMatrix) : Double = {
		val X : ArraySeq[Array[FeatureNode]] = dm.getX();
		val Y : ArraySeq[Double] = dm.getY();
		
	/*	val problem : Problem = new Problem();
		problem.x = X;  // feature nodes
		problem.y = Y;  // target values
		problem.l = X.length; // number of training examples
		problem.n = X(0).length; // number of features
		problem.n = X(0).length; // number of features
	*/
		var correct : Float = 0; var total : Float = 0; var prediction : Double = 0;
		for(i <- 0 until X.length){
//			prediction = Linear.predict(model, X(i));
//			println(dm.get(i));
//			if(prediction == Y(i)) correct = correct + 1; 
			total = total + 1;	
		}		
		correct / total;
	}
	
	def main(args : Array[String]){

		var dm : DataMatrix = new DataMatrix(new ArraySeq[PhrasePair](0));
		dm.initializeFromFile("fake_input_file.txt");
		val (trainDM : DataMatrix, testDM : DataMatrix) = dm splitTrainTest 10;

		train(trainDM);
		println("done");
		//var model : Model = train(trainDM);
		
		//var modelFile : File = new File("model");
		//model save modelFile;
	
		//println(test(model, testDM));
	}
}

File: src/regression/java/Regression.java
import java.io.File;
import java.io.IOException;
import de.bwaldvogel.liblinear.Feature;
import de.bwaldvogel.liblinear.FeatureNode;
import de.bwaldvogel.liblinear.Linear;
import de.bwaldvogel.liblinear.Model;
import de.bwaldvogel.liblinear.Parameter;
import de.bwaldvogel.liblinear.Problem;
import de.bwaldvogel.liblinear.SolverType;

public class Regression {

	/**
 	* Train a model on X with labels Y 
 	*/
	public static Model train(DataMatrix dm){ 
		FeatureNode[][] X = dm.getX();
		double[] Y = dm.getY();
		
		Problem problem = new Problem();
		problem.x = X;  // feature nodes
		problem.y = Y;  // target values
		problem.l = X.length; // number of training examples
		problem.n = X[0].length; // number of features

		SolverType solver = SolverType.L2R_LR; 
		double C = 1.0;    // cost of constraints violation
		double eps = 0.01; // stopping criteria

		Parameter parameter = new Parameter(solver, C, eps);
		return Linear.train(problem, parameter);
	}

	/**
 	* Test model on X with known labels Y 
 	*/
	public static double test(Model model, DataMatrix dm){
		FeatureNode[][] X = dm.getX();
		double[] Y = dm.getY();
		
		Problem problem = new Problem();
		problem.x = X;  // feature nodes
		problem.y = Y;  // target values
		problem.l = X.length; // number of training examples
		problem.n = X[0].length; // number of features

		float correct = 0; float total = 0; double prediction = 0;
		for(int i = 0; i < X.length; i++){
			prediction = Linear.predict(model, X[i]);
			System.out.println(dm.get(i));
			if(prediction == Y[i]){ correct++; }
			total++;			
		}
		return correct / total;

	}
	
	public static void main(String[] args) throws IOException{

		DataMatrix dm = new DataMatrix("fake_input_file.txt");
		DataMatrix[] trainAndTest = dm.splitTrainTest(10);

		DataMatrix trainDM = trainAndTest[0]; 
		DataMatrix testDM = trainAndTest[1];
	
		Model model = train(trainDM);

		File modelFile = new File("model");
		model.save(modelFile);
	
		System.out.println(test(model, testDM));
	}
}
